diff --git a/gdal/port/cpl_error.h b/gdal/port/cpl_error.h
index a90eabd3d27..f75f8899f38 100644
--- a/gdal/port/cpl_error.h
+++ b/gdal/port/cpl_error.h
@@ -130,7 +130,9 @@ typedef int CPLErrorNum;
 /** AWSInvalidCredentials */
 #define CPLE_AWSInvalidCredentials      15
 /** AWSSignatureDoesNotMatch */
-#define CPLE_AWSSignatureDoesNotMatch    16
+#define CPLE_AWSSignatureDoesNotMatch   16
+/** VSIE_AWSError */
+#define CPLE_AWSError                   17
 
 /* 100 - 299 reserved for GDAL */
 
diff --git a/gdal/port/cpl_vsi_error.cpp b/gdal/port/cpl_vsi_error.cpp
index 3c82997acce..cf5c22d6780 100644
--- a/gdal/port/cpl_vsi_error.cpp
+++ b/gdal/port/cpl_vsi_error.cpp
@@ -282,6 +282,10 @@ int CPL_DLL CPL_STDCALL VSIToCPLError( CPLErr eErrClass,
         case VSIE_HttpError:
             CPLError(eErrClass, CPLE_HttpResponse, "%s", VSIGetLastErrorMsg());
             break;
+        case VSIE_AWSError:
+            CPLError(eErrClass, CPLE_AWSError,
+                     "%s", VSIGetLastErrorMsg());
+            break;
         case VSIE_AWSAccessDenied:
             CPLError(eErrClass, CPLE_AWSAccessDenied,
                      "%s", VSIGetLastErrorMsg());
