diff --git a/scripts/ports.cmake b/scripts/ports.cmake
index b3eedf45832074..66b6b5a66939ae 100644
--- a/scripts/ports.cmake
+++ b/scripts/ports.cmake
@@ -234,7 +234,7 @@ elseif(CMD STREQUAL "CREATE")
         message(STATUS "Downloading ${URL} -> ${FILENAME}...")
         file(DOWNLOAD "${URL}" "${DOWNLOAD_PATH}" STATUS download_status)
         list(GET download_status 0 status_code)
-        if(NOT "${download_status}" EQUAL "0")
+        if(NOT "${status_code}" EQUAL "0")
             message(FATAL_ERROR "Downloading ${URL}... Failed. Status: ${download_status}")
         endif()
     endif()
